### PR TITLE
Python2/Python3 compatibility

### PR DESCRIPTION
**Pull request by LegoStormtroopr**
So, I'm close but for some reason on my install image in docs don't show up in python2 and are a little smaller in python3.

I'm gonna fix this even if it kills me.

Todo:

- [x] Figure out how to render a transparent PDF as white (`-flatten` doesn't work for multipage PDFs)
- [ ] Make the images the right size
- [x] Clean up the string.join issues in `reportlab_paragraph`
- [ ] Fix background for tr's

### CODE REVIEW

**Comment by LegoStormtroopr** [comment]
Got the images showing up in Python2 again due to a StringIO/BytesIO issue. Images are still a little smaller than the reference pngs, but they are identical across Python2 and 3 now! :dancer:

**Comment by LegoStormtroopr** [comment]
Todo:

- [ ] Figure out how to render a transparent PDF as white
- [ ] Make the images the right size
- [ ] Clean up the string.join issues in `reportlab_paragraph`

File: tox.ini [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py{26,27,32,33,34}
+envlist = py{27,32,33,34,35}
```
Comment (benjaoming) [inline tox.ini]: Silently removing python 2.6 support? :)

File: tox.ini [hunk 2]
```diff
@@ -41,3 +34,10 @@ basepython = python3.4
 deps =
     reportlab>=3.1,<3.2
     {[testenv]deps}
+
+# === python 3.5 ==============================================================
+[testenv:py26]
```
Comment (benjaoming) [inline tox.ini]: typo I guess

File: xhtml2pdf/xhtml2pdf_reportlab.py [hunk 3]
```diff
@@ -437,6 +436,7 @@ def getRGBData(self):
         return self._data
 
     def getImageData(self):
+        #Is this function ever actually called?
```
Comment (benjaoming) [inline xhtml2pdf/xhtml2pdf_reportlab.py]: it seems like an external interface

File: tox.ini (hunk 1 above)
Comment (benjaoming) [inline tox.ini]: Okay cool, I guess it's because of Reportlab judging from your previous build: https://travis-ci.org/xhtml2pdf/xhtml2pdf/jobs/97829849

Comment (LegoStormtroopr) [inline tox.ini]: Yeah, I know that will ruffle some feathers but Python 2.6.9 was released 2 years ago, and notes with that release state:

> With the 2.6.9 release, and five years after its first release, the Python 2.6 series is now officially retired. All official maintenance for Python 2.6, including security patches, has ended. For ongoing maintenance releases, please see the Python 2.7 series. 

https://www.python.org/download/releases/2.6.9/

File: xhtml2pdf/xhtml2pdf_reportlab.py (hunk 3 above)
Comment (LegoStormtroopr) [inline xhtml2pdf/xhtml2pdf_reportlab.py]: Ok, I'll remove the comment.

**Comment by LegoStormtroopr** [comment]
People should be able to review the current state of rendering here: https://preview.c9users.io/legostormtroopr/py3_xhtml2pdf/testrender/output/index.html

**Comment by LegoStormtroopr** [comment]
It seems that the last stumbling block is that `<tr>`s have an issue with background colors, but that seems to be common in browsers as well, and apart from some flattening issues when converting from PDF to PNG, the lack of support for `<tr>` backgrounds is now the only difference.

Thoughts on how to proceed?

File: tox.ini (hunk 2 above)
Comment (benjaoming) [inline tox.ini]: typo?

**Comment by benjaoming** [comment]
No thoughts on the <tr> issue, sorry :( But great work, hoping to see this merged soon. To wrap it up, you can create a single commit once done using this technique:

http://makandracards.com/makandra/527-squash-several-git-commits-into-a-single-commit

**Comment by LegoStormtroopr** [comment]
@benjaoming Cheers for the link, I have a bad habit of using commit messages as a train of thought. I can clean it up when its more finalised.

It seems that there may be a difference between my version of ImageMagick (6.7.7-10) and that used for the reference images which may be causing some of the differences.

**Comment by LegoStormtroopr** [comment]
Ok, so I have a breakthrough on the `tr` issue

Python 3 is stripping *all* inline background colors rules in tables, but no where else. Python2 is doing it just for `<tr>` cells.

Python2 | Python3
------------ | -------------
Python2 rendering | Python3 rendering

File: tox.ini (hunk 2 above)
Comment (LegoStormtroopr) [inline tox.ini]: typo!

File: xhtml2pdf/xhtml2pdf_reportlab.py [hunk 4]
```diff
@@ -32,10 +32,9 @@
 try:
     import StringIO
 except Exception:
-    from io import StringIO
-    StringIO_old = StringIO
+    from io import BytesIO
     class StringIO(object):
-        StringIO = StringIO_old
+        StringIO = BytesIO
```
Comment (vstoykov) [inline xhtml2pdf/xhtml2pdf_reportlab.py]: I think that here will be better to use:

    try:
        from io import BytesIO
    except ImportError:
        from StringIO import StringIO as BytesIO

And change everywhere in the code `StringIO.StringIO` with `BytesIO`. The `io` module is present even in Python 2.6. The except clause will be only for Python 2.5. Does xhtml2pdf support Python 2.5?

Comment (benjaoming) [inline xhtml2pdf/xhtml2pdf_reportlab.py]: Nope, 2.5 is of the past. If people wanna use that, they should also use old versions of xhtml2pdf :)

Comment (LegoStormtroopr) [inline xhtml2pdf/xhtml2pdf_reportlab.py]: Agreed, 2.5 and 2.6 are no longer supported by Python. This branch is meant to just get Py3 support with no new features.

After this is in, if people want new features they really should upgrade.

**Comment by benjaoming** [comment]
@LegoStormtroopr are you still working on this? :)

I would also like to know if you think I should close #222 because it seems from your comments you started out with the original work from that PR?

**Comment by LegoStormtroopr** [comment]
Hey, thanks for the pull on this. It had a few quirks that I just wanted confirmation on. If there are issues, just assign blame back this way and I'll clean them up.

Hopefully this gets us a little closer to a Py3 build. Thanks for all your hard work!

**Comment by benjaoming** [comment]
I feel like there's a pretty clean slate now, as long as people make PRs against the 'develop' I should manage to merge things easily (I messed up by merging some things into develop and others into master, so it took quite some time to clean up).

Please don't hesitate to deploy the current develop branch into whatever you're working on and feel assured that fixes get merged quickly.

**Comment by benjaoming** [comment]
Btw did the layout issues go away? I found an issue regarding comparison and implemented `__eq__` and `__lt__` because Python3 doesn't use `__cmp__`:

https://github.com/xhtml2pdf/xhtml2pdf/blob/develop/xhtml2pdf/w3c/css.py#L297

**Comment by LegoStormtroopr** [comment]
i never did get this resolved, but thanks for accepting :)

**Comment by benjaoming** [comment]
Your contributions would be very warmly welcomed!! The Python3 stuff that you've worked on really took the project forwards, it was seeing such efforts going to waste that made me take up maintainership.

Btw there's still the issue that you identified regarding background table colors, I created that as a separate issue after discovering it on my own :) #219

**Pull request merged**